xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -300,15 +412,92 @@ public void run() {
     }
   }
 
-  // https://github.com/google/error-prone/issues/1767
-  @SuppressWarnings("ModifyCollectionInEnhancedForLoop")
-  private class ConfigWatcherImpl implements ConfigWatcher {
+  private class ResolveState implements LdsResourceWatcher {
+    private final ConfigOrError emptyServiceConfig =
+        serviceConfigParser.parseServiceConfig(Collections.<String, Object>emptyMap());
+    private final ResolutionResult emptyResult =
+        ResolutionResult.newBuilder()
+            .setServiceConfig(emptyServiceConfig)
+            // let channel take action for no config selector
+            .build();
+    private boolean started;
```

Comment:
I don't think this flag is necessary. It's immediately `true` once instantiated.

Reply:
Ah, this was for the same reason as the `rdsWatcher != this` check for RDS watcher. When there is multi-threading involved, it is possible that a callback was scheduled (to be executed later) while this resolver is being shut down. Then when the callback is invoked, it should be ignored.

Reply:
I would recommend add the _checks_ when you are really implementing the race handling, and review them together. This is really based on assumptions of future implementation which I cannot see the whole picture.

Reply:
Okay, eliminated flags used for racing cases for now.